### ISSUE TITLE
[Coverity] Move from a user-supplied l-value reference

### ISSUE THREAD

**Issue by AlexeySachkov** [intel/llvm#20722]
This is reported as CID `535449`, you can find Coverity issues at https://scan.coverity.com/projects/intel-llvm?tab=overview

https://github.com/intel/llvm/blob/3d70234f25f5a9781b4e8ea8eeb1d8e139431b4d/sycl/include/sycl/handler.hpp#L1463-L1476

We accept `Func` argument as a universal reference, meaning that an end user can supply us with an l-value reference:

https://github.com/intel/llvm/blob/3d70234f25f5a9781b4e8ea8eeb1d8e139431b4d/sycl/include/sycl/handler.hpp#L1709-L1711

Which we will then move, thus creating a potentially unexpected side effect. In the example below the second `cout` results in `nulltpr`.

```c++
#include <sycl/sycl.hpp>

struct Functor {
  Functor() : ptr(new int(42)) {}
  Functor(const Functor &other) : ptr(new int) {
    *ptr = *other.ptr;
  }

  Functor(Functor &&other) {
    ptr = other.ptr;
    other.ptr = nullptr;
  }

  void operator()() const {}

  ~Functor() {
    if (ptr)
      delete ptr;
  }

  int *ptr;
};

int main() {

  sycl::queue q;

  Functor f;

  std::cout << "f.ptr = " << f.ptr << std::endl;

  q.submit([&](sycl::handler &cgh) {
    cgh.host_task(f);
  });

  std::cout << "f.ptr = " << f.ptr << std::endl;

  return 0;
}
```

The spec does not seem to be clear about whether or not it is valid for the implementation to leave the input object in an unspecified state.
Related discussions:
- https://github.com/KhronosGroup/SYCL-Docs/pull/644#discussion_r2363630744
- https://github.com/KhronosGroup/SYCL-Docs/pull/922#discussion_r2421758184

Tagging @aelovikov-intel for awareness.

Without the spec clarification we should probably  avoid doing `std::move`s, switch to `std::forward` and provide a library entry point which accepts a regular reference (because today we only provide one for the universal reference).

**Comment by aelovikov-intel** [comment]
> The spec does not seem to be clear about whether or not it is valid for the implementation to leave the input object in an unspecified state.

I think perfect forwarding in the API implies that `std::forward` will be used on that parameter. E.g., [`std::apply`](https://en.cppreference.com/w/cpp/utility/apply.html) has:

> Given the exposition-only function apply-impl defined as follows:
> ...
> The effect is equivalent to:
> `return apply-impl(std::forward<F>(f), ...);`

As such, we should just change `move` to `forward`.

As for

> provide a library entry point which accepts a regular reference (because today we only provide one for the universal reference)

can you clarify why?
